File: 0039-combination-sum/0039-combination-sum.cpp
class Solution {
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<vector<int>>result;
        vector<int>temp;
        sort(candidates.begin(),candidates.end());
        int n=candidates.size();
        getAllCombinationSum(0,0,n,target,candidates,temp,result);
        return result;
    }
    void getAllCombinationSum(int idx,int sum,int n,int target,vector<int>&arr,vector<int>temp,vector<vector<int>>&result){
        if(idx==n){
            if(sum==target){
                result.push_back(temp);
            }
            return;
        }
        if(target>sum){
            sum+=arr[idx];
            temp.push_back(arr[idx]);
            getAllCombinationSum(idx,sum,n,target,arr,temp,result);
            sum-=arr[idx];
            temp.pop_back();
        }
        getAllCombinationSum(idx+1,sum,n,target,arr,temp,result);
    }
};

File: 0039-combination-sum/0039-combination-sum.java
class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> result=new ArrayList<>();
        Arrays.sort(candidates);
        int n=candidates.length;
        getAllCombinationSum(0,n,target,candidates,new ArrayList<>(),result);
        return result;
    }

    private void getAllCombinationSum(int idx, int n, int target, int[] arr, List<Integer> temp, List<List<Integer>> result) {
        if(target<0){
            return;
        }
        if(idx==n){
            if(target==0){
                result.add(new ArrayList<>(temp));
            }
            return ;
        }
         if(target-arr[idx]>=0){
            temp.add(arr[idx]);
            getAllCombinationSum(idx,n,target-arr[idx],arr,temp,result);
            temp.remove(temp.size()-1);
        }
        getAllCombinationSum(idx+1,n,target,arr,temp,result);
      
    }
}